src/API/QueryFieldService.php

```diff
@@ -48,29 +49,33 @@ public function __construct(
         $this->queryTypeRegistry = $queryTypeRegistry;
     }
 
-    /**
-     * @param \eZ\Publish\API\Repository\Values\Content\Content $content
-     * @param string $fieldDefinitionIdentifier
-     *
-     * @return \eZ\Publish\API\Repository\Values\Content\Content[]
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
-     */
     public function loadContentItems(Content $content, string $fieldDefinitionIdentifier): iterable
     {
-        $query = $this->prepareQuery($content, $fieldDefinitionIdentifier);
+        $query = $this->prepareQuery($content, $content->contentInfo->getMainLocation(), $fieldDefinitionIdentifier);
 
-        return array_map(
-            function (SearchHit $searchHit) {
-                return $searchHit->valueObject;
-            },
-            $this->searchService->findContent($query)->searchHits
-        );
+        return $this->executeQueryAndMapResult($query);
+    }
+
+    public function loadContentItemsForLocation(Location $location, string $fieldDefinitionIdentifier): iterable
+    {
+        $query = $this->prepareQuery($location->getContent(), $location, $fieldDefinitionIdentifier);
+
+        return $this->executeQueryAndMapResult($query);
     }
 
     public function countContentItems(Content $content, string $fieldDefinitionIdentifier): int
     {
-        $query = $this->prepareQuery($content, $fieldDefinitionIdentifier);
+        $query = $this->prepareQuery($content, $content->contentInfo->getMainLocation(), $fieldDefinitionIdentifier);
+        $query->limit = 0;
+
+        $count = $this->searchService->findContent($query)->totalCount - $query->offset;
+
+        return $count < 0 ? 0 : $count;
```

Comment:
Why do you need this magic? What does it count? `totalCount` should always be a correct result.

Reply:
It's not magic, it's offset manipulation. On the last page, the query will most likely return _less_ than $offset results. In that case, we return 0 instead of a negative value.

Reply:
> It's not magic, it's offset manipulation.

In this case, it's highly unintuitive. 
Why even `offset` takes part in the method that's supposed to return total count of objects (`countContentItems`)?

Reply:
Actually the use-case of $offset in that case is quite clear: it is a _hardcoded_ offset that comes from the query type / query field itself (sorry for the misleading description above, it's been a while...).

Example:
- `blog` content type with `blog_post` sub-items
- `highlighted_post_1` query field that returns the most recent post (offset = 0, limit = 1)
- `highlighted_post_2` query field that returns the 2nd most recent post (offset = 1, limit = 1)
- `posts` field that returns the fields _after_ the two highlighted ones (offset = 2)

When counting the results for `posts`, if there are 0 or 1 posts in the folder, count - offset will be -2 or -1.

Reply:
It isn't immediately obvious for someone who is just looking at the PR itself, without remembering all the quirks that `QueryField` type can have.

After the explanation it makes sense. Does any test will prevent this from being accidentally changed if someone like me comes and tries to remove it, thinking it's an error?

Reply:
> After the explanation it makes sense. Does any test will prevent this from being accidentally changed if someone like me comes and tries to remove it, thinking it's an error?

As a matter of fact there is ! The spec is called `it_returns_zero_if_offset_is_bigger_than_count`.